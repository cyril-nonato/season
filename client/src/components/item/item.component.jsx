import React, {useEffect} from 'react'
import CustomButton from '../custom-button/custom-button.component';
import * as S from './item.styles';

const Item = ({ item, shopCleanUpError, addItemToCart }) => {
  const { name, imageUrl, status, battery, memory, price } = item;

  useEffect(() => {
    return () => {
      shopCleanUpError()
    }
  });

  return (
    <S.Container>
      <S.H3 data-testid='h3'>{name}</S.H3>
      <S.Grid>
        <S.Image data-testid='image' src={imageUrl} alt={name} />
        <S.Status>
          <S.Description>Status:</S.Description>
          <S.Value>{status}</S.Value>
        </S.Status>
        <S.Battery>
          <S.Description>Battery:</S.Description>
          <S.Value>{battery}</S.Value>
        </S.Battery>
        <S.Memory>
          <S.Description>Memory:</S.Description>
          <S.Value>{memory}</S.Value>
        </S.Memory>
        <S.Price>
          <S.Description>Price:</S.Description>
          <S.Value>${price.toFixed(2)}</S.Value>
        </S.Price>
        <S.ButtonContainer>
          <CustomButton type='button' click={() => addItemToCart(item)}>Add to cart</CustomButton>
        </S.ButtonContainer>

      </S.Grid>
    </S.Container>
  );
}

export default Item;


/*

        {
          !showComment &&
          <S.LoadComments onClick={() => getCommentsStart(brandName, model)}>load comments</S.LoadComments>
        }
        {
          showComment &&
          <S.CommentsContainer>
            <ItemCommentsContainer />
          </S.CommentsContainer>
        }

*/