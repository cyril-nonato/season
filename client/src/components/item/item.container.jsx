import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { selectItemIsLoaded, selectShopError, selectItem } from '../../redux/shop/shop.selector';
import { selectShowComments } from '../../redux/comments/comments.selector';
import { getCommentsStart } from '../../redux/comments/comments.actions';

import WithSpinner from '../with-spinner/with-spinner.component';
import Item from './item.component';
import { shopCleanUpError } from '../../redux/shop/shop.actions';
import { addItemToCart } from '../../redux/cart/cart.actions';

const mapStateToProps = createStructuredSelector({
  loading: (state) => !selectItemIsLoaded(state),
  error: selectShopError,
  item: selectItem,
  showComment: selectShowComments
});

const mapDispatchToProps = dispatch => ({
  shopCleanUpError: () => dispatch(shopCleanUpError()),
  addItemToCart: (item) => dispatch(addItemToCart(item)),
  getCommentsStart: () => dispatch(getCommentsStart())
})

const ItemContainer = compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps), 
  WithSpinner)(Item)

export default ItemContainer;