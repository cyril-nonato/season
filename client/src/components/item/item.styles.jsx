import styled, { css } from 'styled-components'
import { mediaQueries } from '../../theme/theme';

export const Container = styled.div`
  width: 50%;
  margin: 0 auto;

  @media only screen and (max-width: ${mediaQueries.windows}) {
    width: 60%;
  }

  @media only screen and (max-width: ${mediaQueries.tablet}) {
    width: 80%;
  }

  @media only screen and (max-width: ${mediaQueries.phone}) {
    width: 95%;
  }
`;

const contextStyles = css`
  display: grid;
  grid-template-columns: 1fr 2fr;
  grid-auto-rows: 100%;
  justify-content: center;
  align-items: center;
`;

export const Grid = styled.div`
  margin-top: 2rem;
  display: grid;
  grid-template-columns: 40% 60%;
  grid-template-rows: repeat(5, 8rem) 2rem;
`;

export const Image = styled.img`
  height: 100%;
  grid-row: 1 / 5;
  grid-column: 1 / 2;
  justify-self: center;
`;

export const Status = styled.div`
  grid-row: 1 / 2;
  grid-column: 2 / 3;
  ${contextStyles}
`;

export const Memory = styled.div`
  grid-row: 2 / 3;
  grid-column: 2 / 3;
  ${contextStyles}
`;

export const Battery = styled.div`
  grid-row: 3 / 4;
  grid-column: 2 / 3;
  ${contextStyles}
`;

export const Price = styled.div`
  grid-row: 4 / 5;
  grid-column: 2 / 3;
  ${contextStyles}
`;

export const ButtonContainer = styled.div`
  grid-row: 5 / 6;
  grid-column: 2 / 3;
  justify-self: center;
  align-self: center;

  button {
    padding: 1.5rem 2.5rem;
  }
`

export const Description = styled.span`
  font-weight: bold;
  font-size: 1.4rem;
  text-transform: uppercase;
`;

export const Value = styled.span`
  font-size: 1.6rem;
`;

export const H3 = styled.h3`
  font-size: 1.8rem;
  text-transform: uppercase;
  letter-spacing: 1px;
  padding-bottom: 1rem;
  border-bottom: .2rem solid black;
`;

export const LoadComments = styled.p`
  grid-row: 6 / 7;
  grid-column: 1 / 3;

  text-align: center;
  font-size: 1.6rem;
  text-transform: uppercase;

  &:hover {
    cursor: pointer;
  }
`;

export const CommentsContainer = styled.div`
  grid-row: 6 / 7;
  grid-column: 1 / 3;
`;


// export const StarContainer = styled.div`
//   grid-row: 5 / 6;
//   grid-column: 1 / 2;

//   width: 100%;



// `

// export const Star = styled.img`
//   height: 100%;
//   color: black;
//   z-index: -1;
// `;

export const GridComment = styled.div`

`