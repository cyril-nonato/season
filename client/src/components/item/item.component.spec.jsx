import React from 'react'
import {render, cleanup} from '@testing-library/react'
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import * as S from './item.styles'

import Item from './item.component'

afterEach(cleanup);

describe('Item Component Test', () => {
  const item = {
    name: 'name', 
    imageUrl: 'url', 
    status: 'release', 
    battery: '550', 
    memory: 'some-gb', 
    price: 1000
  }
  const shopCleanupError =  jest.fn()

  let wrapper;

  beforeEach(() => {
    wrapper = render(<Item item={item} shopCleanUpError={shopCleanupError} />)
  })

  test('Should render props correctly', () => {
    const {getByTestId} = wrapper;

    expect(getByTestId('h3')).toHaveTextContent(item.name);
    expect(getByTestId('image')).toHaveAttribute('src', item.imageUrl)
  });

  test('Should call shopCleanupError on unmount ', () => {
    const {unmount} = wrapper;
    unmount();
    expect(shopCleanupError).toHaveBeenCalled();
  });
});

describe('Item Styles Test', () => {
  let tree;
  afterEach(() => {
    expect(tree).toMatchSnapshot();
  })
  test('Container should match the snapshot ', () => {
    tree = renderer.create(<S.Container />).toJSON();
  });

  test('Grid should match the snapshot ', () => {
    tree = renderer.create(<S.Grid />).toJSON();
  });

  test('Image should match the snapshot ', () => {
    tree = renderer.create(<S.Image />).toJSON();
  });

  test('Status should match the snapshot ', () => {
    tree = renderer.create(<S.Status />).toJSON();
  });
});