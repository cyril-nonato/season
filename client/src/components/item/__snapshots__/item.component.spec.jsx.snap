// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Styles Test Container should match the snapshot  1`] = `
.c0 {
  width: 100%;
  margin: 0 auto;
}

@media screen and (min-width:700px) {
  .c0 {
    width: 80%;
  }
}

@media screen and (min-width:900px) {
  .c0 {
    width: 50%;
  }
}

<div
  className="c0"
/>
`;

exports[`Item Styles Test Grid should match the snapshot  1`] = `
.c0 {
  margin-top: 2rem;
  display: grid;
  grid-template-columns: 40% 60%;
  grid-template-rows: repeat(5,8rem) 2rem;
}

<div
  className="c0"
/>
`;

exports[`Item Styles Test Image should match the snapshot  1`] = `
.c0 {
  height: 100%;
  grid-row: 1 / 5;
  grid-column: 1 / 2;
  justify-self: center;
}

<img
  className="c0"
/>
`;

exports[`Item Styles Test Status should match the snapshot  1`] = `
.c0 {
  grid-row: 1 / 2;
  grid-column: 2 / 3;
  display: grid;
  grid-template-columns: 1fr 2fr;
  grid-auto-rows: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="c0"
/>
`;
